# PART 1 
## Installing and Importing programs to read and parse html document

In [1]:
import requests

In [2]:
!pip install BeautifulSoup4
!pip install lxml

#### Passing website which contains a list of neighboorhoods, boroughs in Canada and their postal codes 

In [80]:
website_url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
from bs4 import BeautifulSoup
soup = BeautifulSoup(website_url,'lxml')
#soup.find("table",{"class":"wikitable"})


<table class="wikitable sortable">
<tbody><tr>
<th>Postal code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park / Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor / Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park / Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern / Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>Ea

#### Using tag 'td' to find and split values from the website. 
#### Assign values to the empty lists (i,e., code_list)
#### Grouping information into a table using pandas

In [5]:
import pandas as pd
codes_list=[]
borough_list=[]
neighborhood_list=[]
i=1
for tag in soup.table.find_all('td'):
    if i == 1:
        tag = tag.text
        tag = tag.replace('\n', '')
        codes_list.append(tag)
        
    if i == 2:
        tag = tag.text
        tag = tag.replace('\n', '')
        borough_list.append(tag)
    if i == 3: 
        tag = tag.text
        tag = tag.replace('\n', '')
        neighborhood_list.append(tag)
    i = i+1
    if i==4:
        i=1
#combining lists into dataframe
df = pd.DataFrame()
df['PostalCode']=codes_list
df['Borough']=borough_list 
df['Neighborhood']=neighborhood_list
#replacing / with ,
df=df.replace({'/':','},regex=True)
df

PostalCode           Borough  \
0          M1A      Not assigned   
1          M2A      Not assigned   
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
..         ...               ...   
175        M5Z      Not assigned   
176        M6Z      Not assigned   
177        M7Z      Not assigned   
178        M8Z         Etobicoke   
179        M9Z      Not assigned   

                                          Neighborhood  
0                                                       
1                                                       
2                                            Parkwoods  
3                                     Victoria Village  
4                           Regent Park , Harbourfront  
..                                                 ...  
175                                                     
176                                                     
177                                                     
178  Mimico NW , The Queensway West , South of Bloo...  
179                                                     

[180 rows x 3 columns]

#### Dropping not assigned Borough and reseting index

In [6]:
df = df[df['Borough'] != 'Not assigned'].reset_index(drop=True)
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park , Harbourfront  
3                   Lawrence Manor , Lawrence Heights  
4        Queen's Park , Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern , Rouge  
7                                           Don Mills  
8                    Parkview Hill , Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park , Princess Gardens , Martin Gr...

In [7]:
df.shape

(103, 3)

#### The table above contains 103 neighborhoods

### PART 2

##### installing geocoder to get coordinates of the 103 neighborhoods

In [8]:
!conda install -c conda-forge geocoder --yes

Solving environment: ...working... done

# All requested packages already installed.



In [9]:
import geocoder

#### Obtaning latitute and longitude coordinates of neighborhoods using the PostalCode column

In [10]:
latitude=[]
longitude=[]
# select each row of df 
for i in range(len(df)):
#initialize lat_lng_coords to none
    lat_lng_coords = None    
#loop until you get the coordinates
    while(lat_lng_coords is None):
        g=geocoder.arcgis('{},Toronto, Ontario'.format(df['PostalCode'][i]))
        lat_lng_coords=g.latlng
    latitude.append(lat_lng_coords[0])
    longitude.append(lat_lng_coords[1])
#adding latitude and longitude columns to table
df['Latitude']=latitude
df['Longitude']=longitude


In [11]:
print(df.shape)
df.head()

(103, 5)


PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   

    Latitude  Longitude  
0  43.752935 -79.335641  
1  43.728102 -79.311890  
2  43.650964 -79.353041  
3  43.723265 -79.451211  
4  43.661790 -79.389390

### PART 3

##### Selecting Boroughs that contain the word 'York' in their names

In [12]:
df_York=df[df['Borough'].str.contains('York')].reset_index(drop=True)
print(df_York.shape)
df_York.head(10)


(34, 5)


PostalCode     Borough                       Neighborhood   Latitude  \
0        M3A  North York                          Parkwoods  43.752935   
1        M4A  North York                   Victoria Village  43.728102   
2        M6A  North York  Lawrence Manor , Lawrence Heights  43.723265   
3        M3B  North York                          Don Mills  43.748900   
4        M4B   East York   Parkview Hill , Woodbine Gardens  43.707193   
5        M6B  North York                          Glencairn  43.707279   
6        M3C  North York                          Don Mills  43.722143   
7        M4C   East York                   Woodbine Heights  43.689740   
8        M6C        York                 Humewood-Cedarvale  43.691730   
9        M6E        York                Caledonia-Fairbanks  43.689118   

   Longitude  
0 -79.335641  
1 -79.311890  
2 -79.451211  
3 -79.357220  
4 -79.311529  
5 -79.447500  
6 -79.352023  
7 -79.308507  
8 -79.430013  
9 -79.450650

##### Installing folium needed to visualize neighborhoods in a map

In [14]:
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab


In [114]:
website_url2 = "https://en.wikipedia.org/wiki/Toronto_subway"
table = pd.read_html(website_url2)[2]
table

Line            Stations       Length[1][10]  \
0    Yonge–University                  38   38.8 km (24.1 mi)   
1      Bloor–Danforth                  31   26.2 km (16.3 mi)   
2         Scarborough                   6     6.4 km (4.0 mi)   
3            Sheppard                   5     5.5 km (3.4 mi)   
4  Under construction  Under construction  Under construction   
5            Eglinton                  25       19 km (12 mi)   
6          Finch West                  18      11 km (6.8 mi)   

           Technology  
0          Heavy rail  
1          Heavy rail  
2         Light metro  
3          Heavy rail  
4  Under construction  
5          Light rail  
6          Light rail

In [110]:
page = requests.get("https://en.wikipedia.org/wiki/List_of_volcanoes_by_elevation#From_its_base_on_the_ocean_floor").text
soup3 = BeautifulSoup(page, "lxml")
#tab=soup3.find("table",{"class":"wikitable"})
#print(tab)
print(soup3.title.text)
line=[]
off_peak=[]
rush_hour=[]
e=[]
#f=[]
#g=[]
#h=[]

i=1
for tag in soup3.table.find_all('td'):
    if i == 1:
        tag = tag.text
        tag = tag.replace('\n', '')
        line.append(tag)
        
    if i == 2:
        tag = tag.text
        tag = tag.replace('\n', '')
        off_peak.append(tag)
          
    if i == 3: 
        tag = tag.text
        tag = tag.replace('\n', '')
        rush_hour.append(tag)
    if i == 4: 
        tag = tag.text
        tag = tag.replace('\n', '')
        e.append(tag)
    #if i == 5: 
        #tag = tag.text
        #tag = tag.replace('\n', '')
        #f.append(tag)
    #if i == 6: 
        #tag = tag.text
        #tag = tag.replace('\n', '')
        #g.append(tag)
    #if i == 7: 
        #tag = tag.text
        #tag = tag.replace('\n', '')
        #h.append(tag)
          
    i = i+1
    if i==5:
        i=1
#combining lists into dataframe
df2 = pd.DataFrame()
df2['line_color']=line
df2['offpeak']=off_peak
df2['rushhour']=rush_hour
df2['re']=e
#df2['rf']=f
#df2['rg']=g
#df2['rh']=h
#replacing / with ,
#df2=df2.replace({'/':','},regex=True)
df2

List of volcanoes by elevation - Wikipedia


line_color offpeak rushhour                       re
0  Mauna Kea  10,203   33,474  33474 –  Hawaii, U.S.A

##### Map of all neighborhoods whose Borough contains the word 'York' only

In [15]:

# create map of Manhattan using latitude and longitude values
map_df_York=folium.Map(location=[43.7001114, -79.4162979], zoom_start=10)

# add markers to map
for lat, lng, label in zip(df_York['Latitude'], df_York['Longitude'], df_York['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_df_York)  
    
map_df_York

##### Getting venues 500 meters from the neighborhoods

In [20]:
# @hidden_cell
CLIENT_ID =  'AANQS4UMGB0ZZS43JRAIHXEU0J3BYFFC2RAGGCU0K1XUVYFA' # your Foursquare ID
CLIENT_SECRET = 'JJNSOBEPGUMRJOHRUPS2V4CIOKS0UXPRKSDMWBJSCUXQ1ZDA' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version


##### Defining function getNerbyVenues and using Foursquare API to gather up 100 venues that are 500 meters from each neighboorhood

In [21]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Calling getNerbyVenues and counting the number of unique venue categories retrieved

In [22]:
york_venues = getNearbyVenues(names=df_York['Neighborhood'],
                                   latitudes=df_York['Latitude'],
                                   longitudes=df_York['Longitude']
                                  )
print('\n There are {} uniques categories.'.format(len(york_venues['Venue Category'].unique())))

Parkwoods
Victoria Village
Lawrence Manor , Lawrence Heights
Don Mills
Parkview Hill , Woodbine Gardens
Glencairn
Don Mills
Woodbine Heights
Humewood-Cedarvale
Caledonia-Fairbanks
Leaside
Hillcrest Village
Bathurst Manor , Wilson Heights , Downsview North
Thorncliffe Park
Fairview , Henry Farm , Oriole
Northwood Park , York University
East Toronto
Bayview Village
Downsview
York Mills , Silver Hills
Downsview
North Park , Maple Leaf Park , Upwood Park
Humber Summit
Willowdale , Newtonbrook
Downsview
Bedford Park , Lawrence Manor East
Del Ray , Mount Dennis , Keelsdale and Silverthorn
Humberlea , Emery
Willowdale
Downsview
Runnymede , The Junction North
Weston
York Mills West
Willowdale

 There are 130 uniques categories.


##### 2856 venues were retrieved. There are 130 unique venue categories

In [29]:
print('Total number of venues retrieved: ',york_venues.size)
york_venues.head()

Total number of venues retrieved:  2856


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.752935              -79.335641   
1         Parkwoods              43.752935              -79.335641   
2         Parkwoods              43.752935              -79.335641   
3  Victoria Village              43.728102              -79.311890   
4  Victoria Village              43.728102              -79.311890   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  MacLeod Exteriors Inc.       43.755014       -79.338688   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

               Venue Category  
0                        Park  
1           Food & Drink Shop  
2  Construction & Landscaping  
3                 Coffee Shop  
4       Portuguese Restaurant

##### Rearrenging table to show "Venue Catergories" as headers

In [34]:
york_onehot = pd.get_dummies(york_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
york_onehot['Neighborhood'] = york_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [york_onehot.columns[-1]] + list(york_onehot.columns[:-1])
york_onehot = york_onehot[fixed_columns]

york_onehot.head(7)


Neighborhood  Airport  American Restaurant  Arts & Crafts Store  \
0         Parkwoods        0                    0                    0   
1         Parkwoods        0                    0                    0   
2         Parkwoods        0                    0                    0   
3  Victoria Village        0                    0                    0   
4  Victoria Village        0                    0                    0   
5  Victoria Village        0                    0                    0   
6  Victoria Village        0                    0                    0   

   Athletics & Sports  Auto Dealership  Baby Store  Bagel Shop  Bakery  Bank  \
0                   0                0           0           0       0     0   
1                   0                0           0           0       0     0   
2                   0                0           0           0       0     0   
3                   0                0           0           0       0     0   
4                   0                0           0           0       0     0   
5                   0                0           0           0       0     0   
6                   0                0           0           0       0     0   

   ...  Tea Room  Thai Restaurant  Theater  Toy / Game Store  Trail  \
0  ...         0                0        0                 0      0   
1  ...         0                0        0                 0      0   
2  ...         0                0        0                 0      0   
3  ...         0                0        0                 0      0   
4  ...         0                0        0                 0      0   
5  ...         0                0        0                 0      0   
6  ...         0                0        0                 0      0   

   Video Game Store  Video Store  Vietnamese Restaurant  Women's Store  \
0                 0            0                      0              0   
1                 0            0                      0              0   
2                 0            0                      0              0   
3                 0            0                      0              0   
4                 0            0                      0              0   
5                 0            0                      0              0   
6                 0            0                      0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  
5            0  
6            0  

[7 rows x 131 columns]

##### Grouping rows by neighborhood and by taking the mean of the frequency of occurrence of each venue category

In [39]:
york_grouped = york_onehot.groupby('Neighborhood').mean().reset_index()
york_grouped.head()

Neighborhood  Airport  \
0  Bathurst Manor , Wilson Heights , Downsview North      0.0   
1                                    Bayview Village      0.0   
2                 Bedford Park , Lawrence Manor East      0.0   
3                                Caledonia-Fairbanks      0.0   
4  Del Ray , Mount Dennis , Keelsdale and Silvert...      0.0   

   American Restaurant  Arts & Crafts Store  Athletics & Sports  \
0                  0.0                  0.0                 0.0   
1                  0.0                  0.0                 0.0   
2                  0.0                  0.0                 0.0   
3                  0.0                  0.0                 0.0   
4                  0.0                  0.0                 0.0   

   Auto Dealership  Baby Store  Bagel Shop    Bakery      Bank  ...  Tea Room  \
0              0.0         0.0         0.0  0.000000  0.105263  ...       0.0   
1              0.0         0.0         0.0  0.000000  0.000000  ...       0.0   
2              0.0         0.0         0.0  0.000000  0.000000  ...       0.0   
3              0.0         0.0         0.0  0.111111  0.000000  ...       0.0   
4              0.0         0.0         0.0  0.000000  0.000000  ...       0.0   

   Thai Restaurant  Theater  Toy / Game Store     Trail  Video Game Store  \
0         0.000000      0.0               0.0  0.000000               0.0   
1         0.000000      0.0               0.0  0.333333               0.0   
2         0.045455      0.0               0.0  0.000000               0.0   
3         0.000000      0.0               0.0  0.000000               0.0   
4         0.000000      0.0               0.0  0.000000               0.0   

   Video Store  Vietnamese Restaurant  Women's Store  Yoga Studio  
0     0.052632                    0.0       0.000000          0.0  
1     0.000000                    0.0       0.000000          0.0  
2     0.000000                    0.0       0.000000          0.0  
3     0.000000                    0.0       0.111111          0.0  
4     0.000000                    0.0       0.000000          0.0  

[5 rows x 131 columns]

##### Defining function return_most_common_venues to obtain the top venues for each neighborhood

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

##### Calling function and rearrenging table to show the top 10 venues 

In [45]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = york_grouped['Neighborhood']

for ind in np.arange(york_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(york_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0  Bathurst Manor , Wilson Heights , Downsview North   
1                                    Bayview Village   
2                 Bedford Park , Lawrence Manor East   
3                                Caledonia-Fairbanks   
4  Del Ray , Mount Dennis , Keelsdale and Silvert...   

        1st Most Common Venue 2nd Most Common Venue  \
0                 Coffee Shop                  Bank   
1                 Flower Shop                 Trail   
2              Sandwich Place           Coffee Shop   
3                        Park                   Gym   
4  Construction & Landscaping           Coffee Shop   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Middle Eastern Restaurant              Pharmacy           Pizza Place   
1  Construction & Landscaping                   Gym         Grocery Store   
2          Italian Restaurant      Sushi Restaurant        Cosmetics Shop   
3                      Bakery    Mexican Restaurant                   Spa   
4         Distribution Center                 Field  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Deli / Bodega                 Diner            Restaurant   
1     Convenience Store        Cosmetics Shop         Deli / Bodega   
2            Restaurant                   Pub             Juice Bar   
3            Beer Store   Sporting Goods Shop         Women's Store   
4    Falafel Restaurant     Electronics Store               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0           Bridal Shop         Sandwich Place  
1      Department Store           Dessert Shop  
2               Butcher            Sports Club  
3           Gas Station              Gastropub  
4       Doctor's Office            Yoga Studio

#### Clustering Neighborhoods based on their top 10 venue categories using 3 clusters

In [46]:
from sklearn.cluster import KMeans 
# set number of clusters
kclusters = 3

york_grouped_clustering = york_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(york_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 2, 0, 0, 0, 0, 1])

In [47]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

york_merged = df_York

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
york_merged = york_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood',how='right')

york_merged.head(10) # check the last columns!

PostalCode     Borough                       Neighborhood   Latitude  \
0        M3A  North York                          Parkwoods  43.752935   
1        M4A  North York                   Victoria Village  43.728102   
2        M6A  North York  Lawrence Manor , Lawrence Heights  43.723265   
3        M3B  North York                          Don Mills  43.748900   
6        M3C  North York                          Don Mills  43.722143   
4        M4B   East York   Parkview Hill , Woodbine Gardens  43.707193   
5        M6B  North York                          Glencairn  43.707279   
7        M4C   East York                   Woodbine Heights  43.689740   
8        M6C        York                 Humewood-Cedarvale  43.691730   
9        M6E        York                Caledonia-Fairbanks  43.689118   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.335641               1  Construction & Landscaping   
1 -79.311890               0                 Pizza Place   
2 -79.451211               0              Clothing Store   
3 -79.357220               0          Athletics & Sports   
6 -79.352023               0          Athletics & Sports   
4 -79.311529               0                 Pizza Place   
5 -79.447500               0                 Pizza Place   
7 -79.308507               0                    Pharmacy   
8 -79.430013               0                Hockey Arena   
9 -79.450650               0                        Park   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop                    Park   Distribution Center   
1           Coffee Shop   Portuguese Restaurant                  Park   
2   American Restaurant  Furniture / Home Store            Restaurant   
3                   Gym                    Bank                  Park   
6                   Gym                    Bank                  Park   
4  Fast Food Restaurant                    Café             Pet Store   
5           Gas Station                     Pub   Rental Car Location   
7           Gas Station          Breakfast Spot       Doctor's Office   
8      Business Service           Grocery Store                 Trail   
9                   Gym                  Bakery    Mexican Restaurant   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Fast Food Restaurant    Falafel Restaurant     Electronics Store   
1              Intersection     French Restaurant        Discount Store   
2          Toy / Game Store            Food Court           Men's Store   
3               Coffee Shop  Other Great Outdoors            Restaurant   
6               Coffee Shop  Other Great Outdoors            Restaurant   
4        Athletics & Sports  Gym / Fitness Center             Gastropub   
5  Mediterranean Restaurant   Japanese Restaurant  Fast Food Restaurant   
7                       Bar      Sushi Restaurant         Grocery Store   
8                      Park                 Field    Falafel Restaurant   
9                       Spa            Beer Store   Sporting Goods Shop   

  8th Most Common Venue   9th Most Common Venue      10th Most Common Venue  
0               Dog Run         Doctor's Office                 Yoga Studio  
1     Electronics Store                 Dog Run             Doctor's Office  
2             Bookstore          Cosmetics Shop                    Pharmacy  
3          Burger Joint                     Spa  Construction & Landscaping  
6          Burger Joint                     Spa  Construction & Landscaping  
4    Rock Climbing Spot            Intersection                        Bank  
5         Grocery Store  Furniture / Home Store                   Gift Shop  
7           Coffee Shop     Arts & Crafts Store            Department Store  
8     Electronics Store                 Dog Run             Doctor's Office  
9         Women's Store             Gas Station                   Gastropub

####  Visualize the resulting clusters

In [48]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[43.7001114, -79.4162979], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(york_merged['Latitude'], york_merged['Longitude'], york_merged['Neighborhood'], york_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 0 red color. Neigborhoods like Parks

In [49]:
york_merged.loc[york_merged['Cluster Labels'] == 0, york_merged.columns[[2] + list(range(5, york_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
1                                    Victoria Village               0   
2                   Lawrence Manor , Lawrence Heights               0   
3                                           Don Mills               0   
6                                           Don Mills               0   
4                    Parkview Hill , Woodbine Gardens               0   
5                                           Glencairn               0   
7                                    Woodbine Heights               0   
8                                  Humewood-Cedarvale               0   
9                                 Caledonia-Fairbanks               0   
10                                            Leaside               0   
12  Bathurst Manor , Wilson Heights , Downsview North               0   
13                                   Thorncliffe Park               0   
14                     Fairview , Henry Farm , Oriole               0   
15                   Northwood Park , York University               0   
17                                    Bayview Village               0   
18                                          Downsview               0   
20                                          Downsview               0   
24                                          Downsview               0   
29                                          Downsview               0   
19                          York Mills , Silver Hills               0   
23                           Willowdale , Newtonbrook               0   
25                 Bedford Park , Lawrence Manor East               0   
28                                         Willowdale               0   
33                                         Willowdale               0   
30                     Runnymede , The Junction North               0   
31                                             Weston               0   

     1st Most Common Venue 2nd Most Common Venue       3rd Most Common Venue  \
1              Pizza Place           Coffee Shop       Portuguese Restaurant   
2           Clothing Store   American Restaurant      Furniture / Home Store   
3       Athletics & Sports                   Gym                        Bank   
6       Athletics & Sports                   Gym                        Bank   
4              Pizza Place  Fast Food Restaurant                        Café   
5              Pizza Place           Gas Station                         Pub   
7                 Pharmacy           Gas Station              Breakfast Spot   
8             Hockey Arena      Business Service               Grocery Store   
9                     Park                   Gym                      Bakery   
10     Sporting Goods Shop           Coffee Shop      Furniture / Home Store   
12             Coffee Shop                  Bank   Middle Eastern Restaurant   
13       Indian Restaurant           Yoga Studio                 Supermarket   
14          Clothing Store            Restaurant        Fast Food Restaurant   
15  Furniture / Home Store                   Spa                         Bar   
17             Flower Shop                 Trail  Construction & Landscaping   
18             Coffee Shop  Caribbean Restaurant               Grocery Store   
20             Coffee Shop  Caribbean Restaurant               Grocery Store   
24             Coffee Shop  Caribbean Restaurant               Grocery Store   
29             Coffee Shop  Caribbean Restaurant               Grocery Store   
19                    Bank           Coffee Shop              Cosmetics Shop   
23                    Café           Bridal Shop                 Yoga Studio   
25          Sandwich Place           Coffee Shop          Italian Restaurant   
28             Coffee Shop      Ramen Restaurant                        Park   
33             Coffee Shop      Ramen Restaurant                        Park   
30                 Brewery           Coffee Shop                     

#### Cluster 1 purple color. Neigborhoods like Coffe Shops

In [50]:
york_merged.loc[york_merged['Cluster Labels'] == 1, york_merged.columns[[2] + list(range(5, york_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                    Parkwoods               1   
11                           Hillcrest Village               1   
21  North Park , Maple Leaf Park , Upwood Park               1   
27                           Humberlea , Emery               1   
32                             York Mills West               1   

         1st Most Common Venue                     2nd Most Common Venue  \
0   Construction & Landscaping                         Food & Drink Shop   
11                        Park  Residential Building (Apartment / Condo)   
21                      Bakery                                      Park   
27                  Playground                                      Park   
32                        Park                         Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Park   Distribution Center  Fast Food Restaurant   
11               Dog Run           Yoga Studio                 Field   
21      Basketball Court   Distribution Center  Fast Food Restaurant   
27                 Field     Convenience Store        Cosmetics Shop   
32                  Bank  Fast Food Restaurant    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Falafel Restaurant     Electronics Store               Dog Run   
11        Cosmetics Shop         Deli / Bodega      Department Store   
21    Falafel Restaurant     Electronics Store               Dog Run   
27         Deli / Bodega      Department Store          Dessert Shop   
32     Electronics Store               Dog Run       Doctor's Office   

   9th Most Common Venue 10th Most Common Venue  
0        Doctor's Office            Yoga Studio  
11          Dessert Shop                  Diner  
21       Doctor's Office            Yoga Studio  
27                 Diner         Discount Store  
32   Distribution Center         Discount Store

#### Cluster 2 teal color. Neigborhoods like Construction & Landscaping

In [51]:
york_merged.loc[york_merged['Cluster Labels'] == 2, york_merged.columns[[2] + list(range(5, york_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
26  Del Ray , Mount Dennis , Keelsdale and Silvert...               2   

         1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
26  Construction & Landscaping           Coffee Shop   Distribution Center   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
26                 Field  Fast Food Restaurant    Falafel Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
26     Electronics Store               Dog Run       Doctor's Office   

   10th Most Common Venue  
26            Yoga Studio

In [1]:
from pandas import numpy as np

ImportError: cannot import name 'numpy' from 'pandas' (C:\Users\hp\anaconda3\lib\site-packages\pandas\__init__.py)

In [2]:
import numpy as np

In [4]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])